base/dcp_sharded_cfg.go

```diff
@@ -0,0 +1,146 @@
+package base
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/couchbase/cbgt"
+	"github.com/couchbase/gocb"
+)
+
+// CfgSG is an implementation of Cfg that uses Sync Gateway's existing
+// bucket connection, and existing caching feed to get change notifications.
+//
+type CfgSG struct {
+	bucket        *gocb.Bucket
+	loggingCtx    context.Context
+	subscriptions map[string][]chan<- cbgt.CfgEvent // Keyed by key
+	lock          sync.Mutex                        // mutex for subscriptions
+}
+
+// NewCfgCB returns a Cfg implementation that reads/writes its entries
+// from/to a couchbase bucket, using DCP streams to subscribe to
+// changes.
+//
+// urlStr: single URL or multiple URLs delimited by ';'
+// bucket: couchbase bucket name
+func NewCfgSG(bucket *gocb.Bucket) (*CfgSG, error) {
+
+	cfgContextID := fmt.Sprintf("%s-%s", MD(bucket.Name()), DCPImportFeedID)
+	loggingCtx := context.WithValue(context.Background(), LogContextKey{},
+		LogContext{CorrelationID: cfgContextID},
+	)
+
+	c := &CfgSG{
+		bucket:        bucket,
+		loggingCtx:    loggingCtx,
+		subscriptions: make(map[string][]chan<- cbgt.CfgEvent),
+	}
+	return c, nil
+}
+
+func sgCfgBucketKey(cfgKey string) string {
+	return SGCfgPrefix + cfgKey
+}
+
+func (c *CfgSG) Get(cfgKey string, cas uint64) (
+	[]byte, uint64, error) {
+
+	InfofCtx(c.loggingCtx, KeyDCP, "cfg_sg: Get, key: %s, cas: %d", cfgKey, cas)
+	bucketKey := sgCfgBucketKey(cfgKey)
+	var value []byte
+	casOut, err := c.bucket.Get(bucketKey, &value)
+	if err != nil && err != gocb.ErrKeyNotFound {
+		InfofCtx(c.loggingCtx, KeyDCP, "cfg_sg: Get, ErrKeyNotFound key: %s, cas: %d", cfgKey, cas)
```

Comment:
Not a KeyNotFound error

```suggestion
		InfofCtx(c.loggingCtx, KeyDCP, "cfg_sg: Get Error key: %s, cas: %d err:%s", cfgKey, cas, err)
```

Reply:
Thanks - will fix.